### ISSUE TITLE
SQLAlchemy error during initial import: "does not support in-place multirow inserts"

### ISSUE THREAD

**Issue by jhujhiti** [irrdnet/irrd#503]
**Describe the bug**
While doing initial database imports, I'm seeing a strange error from SQLAlchemy: "sqlalchemy.exc.CompileError: The 'default' dialect with current database version settings does not support in-place multirow inserts." This appears to stop the import in its tracks until the next time the import timer goes off, when it tries and fails again. The full stack trace is:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/irrd/mirroring/mirror_runners_import.py", line 50, in run
    self.full_import_runner.run(database_handler=self.database_handler,
  File "/usr/local/lib/python3.9/site-packages/irrd/mirroring/mirror_runners_import.py", line 216, in run
    p.run_import()
  File "/usr/local/lib/python3.9/site-packages/irrd/mirroring/parsers.py", line 162, in run_import
    self.database_handler.upsert_rpsl_object(
  File "/usr/local/lib/python3.9/site-packages/irrd/storage/database_handler.py", line 215, in upsert_rpsl_object
    self._flush_rpsl_object_writing_buffer()
  File "/usr/local/lib/python3.9/site-packages/irrd/storage/database_handler.py", line 407, in _flush_rpsl_object_writing_buffer
    logger.error(f'Exception occurred while executing statement: {stmt}, rolling back', exc_info=exc)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 486, in __str__
    return str(self.compile())
  File "<string>", line 1, in <lambda>
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 476, in compile
    return self._compiler(dialect, bind=bind, **kw)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 482, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 590, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 319, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 350, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 95, in _compiler_dispatch
    return meth(self, **kw)
  File "/usr/local/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 2444, in visit_insert
    raise exc.CompileError(
sqlalchemy.exc.CompileError: The 'default' dialect with current database version settings does not support in-place multirow inserts.
```

I'm running PostgreSQL 13, but I've tested against 10 with the same result. Judging from the SQLAlchemy code (with which I am not familiar), it seems like this happens before talking to the database at all.

**To Reproduce**
This only seems to occur when I run an initial import with the following configuration:

```yaml
irrd:
  database_url: postgresql://dbserver/db
  redis_url: unix:///shared/redis.sock
  piddir: /var/run/
  server:
    whois:
      interface: "::"
      port: 43
      max_connections: 10
  email:
    from: sanitized@mycompany.com
    smtp: localhost
  rpki:
    roa_source: null
  auth:
    gnupg_keyring: /tmp/asdf.gpg
  sources_default:
    - RADB
    - ARIN
    - RIPE
    - APNIC
    - LACNIC
    - ALTDB
  sources:
    RADB:
      nrtm_host: whois.radb.net
      import_source: ftp://ftp.radb.net/radb/dbase/radb.db.gz
      import_serial_source: ftp://ftp.radb.net/radb/dbase/RADB.CURRENTSERIAL
      import_timer: 3600
      rpki_excluded: true
    ARIN:
      nrtm_host: whois.arin.net
      import_source: ftp://ftp.arin.net/pub/rr/arin.db.gz
      import_serial_source: https://ftp.arin.net/pub/rr/ARIN.CURRENTSERIAL
      import_timer: 3600
      rpki_excluded: true
    RIPE:
      nrtm_host: whois.ripe.net
      nrtm_port: 4444
      import_source: https://ftp.ripe.net/ripe/dbase/ripe.db.gz
      import_serial_source: https://ftp.ripe.net/ripe/dbase/RIPE.CURRENTSERIAL
      import_timer: 3601
      rpki_excluded: true
      object_class_filter:
        - as-set
        - aut-num
        - filter-set
        - inet-rtr
        - key-cert
        - mntner
        - peering-set
        - route
        - route6
        - route-set
        - rtr-set
    APNIC:
      nrtm_host: whois.apnic.net
      import_source:
        - https://ftp.apnic.net/apnic/whois/apnic.db.route.gz
        - https://ftp.apnic.net/apnic/whois/apnic.db.role.gz
        - https://ftp.apnic.net/apnic/whois/apnic.db.route6.gz
        - https://ftp.apnic.net/apnic/whois/apnic.db.aut-num.gz
        - https://ftp.apnic.net/apnic/whois/apnic.db.mntner.gz
        - https://ftp.apnic.net/apnic/whois/apnic.db.as-set.gz
        - https://ftp.apnic.net/apnic/whois/apnic.db.route-set.gz
      import_serial_source: https://ftp.apnic.net/apnic/whois/APNIC.CURRENTSERIAL
      import_timer: 3600
      rpki_excluded: true
    LACNIC:
      nrtm_host: whois.lacnic.net
      import_source: http://irr.lacnic.net/lacnic.db.gz
      import_serial_source: http://irr.lacnic.net/LACNIC.CURRENTSERIAL
      import_timer: 3600
      rpki_excluded: true
    ALTDB:
      nrtm_host: whois.altdb.net
      import_source: ftp://ftp.altdb.net/pub/altdb/altdb.db.gz
      import_serial_source: ftp://ftp.altdb.net/pub/altdb/ALTDB.CURRENTSERIAL
      import_timer: 3600
      rpki_excluded: true
```

I've experimented and found that I cannot reproduce it if I remove RIPE from the configuration during the initial imports, and it works fine when I add RIPE back as it is there, once the imports of the other databases are complete. But starting irrd for the first time with that configuration fails reliably.

**Expected behaviour**
The import should work :)

**IRRd version you are running**
4.1.2 (installed from PyPI with pip) on Python 3.9. I've confirmed that pip did correctly install SQLAlchemy 1.3.18 rather than the latest version.

**Additional context**
My PostgreSQL connection is running through pgpool to an active/passive replication setup, but that's probably not relevant since the error happens during SQLAlchemy compilation rather than transaction execution.

I initially built this configuration one database at a time, restarting after each import, and had no issues. It was only once I built a production environment and reran the imports in bulk that it appeared.

**Comment by mxsasha** [comment]
Thanks for the detailed report! My quick reading is also that this particular error occurs before running the ongoing INSERT query on the database. I suspect there is an earlier error that caused SQLAlchemy to end up in a state where it can not process the ongoing query. Do you see any earlier errors in the IRRd or PostgreSQL log?

(Note that it is normal for IRRd to log some errors about objects that could not be parsed, especially during a full insert, with "Parsing errors occurred while importing from file for ...". This is a consequence of some sources still having unparsable objects. )

**Comment by jhujhiti** [comment]
You're exactly right. I found the issue! Thanks. Bit embarrassed to have opened this issue because the extremely long lines are confusing my logging system and overflowing my huge terminal scrollback window, but it was ultimately caused by a Postgres worker process OOMing. I'm even more confused by the SQLAlchemy code path that results in a different database dialect than the one that irrd explicitly selected, but that doesn't matter to me anymore. For anybody who finds this in the future, here's what it looks like:

irrd logs:
```
irrd[10]: [irrd.storage.preload#CRITICAL] Updating preload store failed, retrying in 5s, traceback follows: (psycopg2.errors.InternalError_) unable to read data from DB node 0
DETAIL:  EOF encountered with backend
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```
is printed immediately before SQLAlchemy prints the contents of what it was trying to insert (which happens to be absolutely enormous). This error was difficult to find until I filtered out those log messages with insert contents.

Postgres logs `LOG:  server process (PID 13465) was terminated by signal 9: Killed` shortly before this. My deployment is running in Kubernetes with resource limits, and the memory consumption happened fast enough that it didn't show up on graphs and Postgres is resilient enough to not crash entirely, so K8s thought everything was fine.
